src/main/java/redis/clients/jedis/JedisClusterInfoCache.java

```diff
@@ -79,36 +80,41 @@ public void discoverClusterNodesAndSlots(Jedis jedis) {
   }
 
   public void discoverClusterSlots(Jedis jedis) {
-    w.lock();
+    //If rediscovering is already in process - no need to start one more same rediscovering, just return
+    if (!rediscovering) {
+      w.lock();
+      rediscovering = true;
```

Comment:
Ah, right. This _looks_ like a candidate for a double-checked locking pattern, but because we're resetting `rediscovering` it won't work. Darn. It'd be nice to be able to get rid of those extra rediscovers.

(And this, friends, is why I shouldn't read PRs before coffee.)